use-ai part updated

yunks128 opened the pull request:
## Purpose
The purpose of this PR is to enhance the flexibility and usability of the AI-based best practice application in the SLIM repository. This update introduces support for multiple AI models, including Ollama and Azure, and adds the ability to specify the model to be used for generating content based on best practices. It also improves the integration of AI-generated content into the workflow and updates the CLI to support these new options.

## Proposed Changes
- **[ADD]** Support for multiple AI models (`Ollama`,  `Azure`, `Openai`) in the SLIM repository.
- **[ADD]** New parameter `model` in functions `apply_best_practice`, `apply_best_practices`, and `apply_and_deploy_best_practices` to specify the AI model to be used.
- **[CHANGE]** Updated the SLIM CLI (`create_parser`) to accept `--model` as an argument, enabling users to select the AI model (`ollama/llama3.1`, `openai/gpt-4o`, `azure/gpt-4o`,etc.) for best practice application.
- **[CHANGE]** Refactored the AI content generation logic to handle multiple providers (`OpenAI`, `Azure`, `Ollama`) and to improve the error handling during content generation.
- **[CHANGE]** Modified existing functions to pass the selected model to the AI content generation process.
- **[FIX]** Addressed potential issues with error handling during AI content generation for Azure and Ollama models.

## Issues
- Resolves issues related to AI model flexibility and provider integration in the SLIM repository. https://github.com/NASA-AMMOS/slim-cli/issues/3

## Testing
- Verified the integration of multiple AI models (`Openai`, `Ollama`) by running tests on supported platforms.
- Tested on Linux and macOS to ensure compatibility.

riverma commented on slim-cli.py:
```diff
@@ -197,14 +214,59 @@ def read_file_content(file_path: str) -> Optional[str]:
 
+import os
+import requests
+from azure.identity import ClientSecretCredential
```
@yunks128 can you add this new module to `requirements.txt`? Note you'll have to be specific about the module from Azure. I got the below when trying to add `azure` to `requirements.txt`:

```
Collecting azure (from -r requirements.txt (line 7))
  Downloading azure-5.0.0.zip (4.6 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [25 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/xm/g2pq18gx5gbdlydq5s84qxnm0000gq/T/pip-install-jwuqe699/azure_f47cd1b34eb64c698d61fd0820a7dd35/setup.py", line 60, in <module>
          raise RuntimeError(message)
      RuntimeError:
      
      Starting with v5.0.0, the 'azure' meta-package is deprecated and cannot be installed anymore.
      Please install the service specific packages prefixed by `azure` needed for your application.
      
      The complete list of available packages can be found at:
      https://aka.ms/azsdk/python/all
      
      Here's a non-exhaustive list of common packages:
      
      -  azure-mgmt-compute (https://pypi.python.org/pypi/azure-mgmt-compute) : Management of Virtual Machines, etc.
      -  azure-mgmt-storage (https://pypi.python.org/pypi/azure-mgmt-storage) : Management of storage accounts.
      -  azure-mgmt-resource (https://pypi.python.org/pypi/azure-mgmt-resource) : Generic package about Azure Resource Management (ARM)
      -  azure-keyvault-secrets (https://pypi.python.org/pypi/azure-keyvault-secrets) : Access to secrets in Key Vault
      -  azure-storage-blob (https://pypi.python.org/pypi/azure-storage-blob) : Access to blobs in storage accounts
      
      A more comprehensive discussion of the rationale for this decision can be found in the following issue:
      https://github.com/Azure/azure-sdk-for-python/issues/10646
      
      
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed
```

riverma replied:
Resolved in df8eb86

riverma commented on slim-cli.py:
```diff
@@ -197,14 +214,59 @@ def read_file_content(file_path: str) -> Optional[str]:
         logging.error(f"Error reading file {file_path}: {e}")
         return None
 
+import os
```
These imports should all be at the top of the file, not in the middle of the code.

riverma commented on slim-cli.py:
```diff
@@ -197,14 +214,59 @@ def read_file_content(file_path: str) -> Optional[str]:
+DEPLOYMENT_ID = os.getenv("DEPLOYMENT_ID")
+
+authority = "https://login.microsoftonline.com"
+credential = ClientSecretCredential(
```
Here - you're initializing Azure services even though no arguments to slim-cli have been passed yet. You need to only initialize Azure services if someone picks that model.

e.g. I'm currently getting this error without any arguments:
```
% python slim-cli.py                            
Traceback (most recent call last):
  File "slim-cli.py", line 231, in <module>
    credential = ClientSecretCredential(
  File "/venv/lib/python3.9/site-packages/azure/identity/_credentials/client_secret.py", line 45, in __init__
    raise ValueError("client_id should be the id of a Microsoft Entra application")
ValueError: client_id should be the id of a Microsoft Entra application
```

riverma commented on slim-cli.py:
```diff
@@ -197,14 +214,59 @@ def read_file_content(file_path: str) -> Optional[str]:
+
+load_dotenv()
+
+TENANT_ID = os.getenv("AZURE_TENANT_ID")
```
Only load / check for specific environment variables if Azure services are being used or loaded. Also, please check and ensure you're giving the user an error message if needed environment variables are not set.

riverma commented on slim-cli.py:
```diff
@@ -259,7 +321,7 @@ def generate_git_branch_name(best_practice_ids):
         return None
 
-def apply_best_practices(best_practice_ids, use_ai_flag, repo_urls = None, existing_repo_dir = None, target_dir_to_clone_to = None):
+def apply_best_practices(best_practice_ids, use_ai_flag, model, repo_urls = None, existing_repo_dir = None, target_dir_to_clone_to = None):
```
Have a default model available

riverma commented on slim-cli.py:
```diff
@@ -682,14 +749,15 @@ def apply_and_deploy_best_practices(best_practice_ids, use_ai_flag, remote_name
             if git_repo:
                 deploy_best_practice(
-                    best_practice_id=best_practice_id,
+                    best_practice_id=best_practice_ids[0],
```
Same as line 750

riverma commented on slim-cli.py:
```diff
@@ -776,11 +847,13 @@ def create_parser():
     parser_apply_deploy.add_argument('--repo-dir', required=False, help='Repository directory location on local machine. Only one repository supported')
     parser_apply_deploy.add_argument('--clone-to-dir', required=False, help='Local path to clone repository to. Compatible with --repo-urls')
     parser_apply_deploy.add_argument('--use-ai', action='store_true', help='Automatically customize the application of the best practice')
+    parser_apply_deploy.add_argument('--model', required=False, help='Model name (ollama/gpt-4o) for using ai')
```
I think it's better to just pass the model name within the `--use-ai` argument rather than have another `--model`. For example: `--use-ai ollama/llama3.1`. What does having two arguments gain? It would only make sense to separate them if you can use the independently of each other, but that's not the case right?

Also the "help" field  `ollama/gpt-4o` needs to be changed since OpenAI is not going to give their model away as open source ;)

riverma commented:
On line https://github.com/NASA-AMMOS/slim-cli/pull/4/files#diff-44f8b6300d8a6dea02d52438f4024065de89201886308402330018a2e531bdfeR285 - you might want to consider not using a Unix process, but rather using https://github.com/ollama/ollama-python. It's cleaner code to maintain.

yunks128 commented:
> On line
> 
> https://github.com/NASA-AMMOS/slim-cli/blob/df8eb86f477728f788a703be584c73d4128f3eb5/slim-cli.py#L285
> 
> - you might want to consider not using a Unix process, but rather using https://github.com/ollama/ollama-python. It's cleaner code to maintain.

@riverma Thank you for reviewing my update! I addressed all your suggestions!
